### PR TITLE
Switch file watcher to stop use polling; bump vscode to 1.66.0

### PR DESCRIPTION
**Pull request by pokey**
- Fixes https://github.com/cursorless-dev/cursorless/issues/1808
- Fixes https://github.com/cursorless-dev/cursorless/issues/1846

## Checklist

- [-] I have added [tests](https://www.cursorless.org/docs/contributing/test-case-recorder/)
- [-] I have updated the [docs](https://github.com/cursorless-dev/cursorless/tree/main/docs) and [cheatsheet](https://github.com/cursorless-dev/cursorless/tree/main/cursorless-talon/src/cheatsheet)
- [-] I have not broken the cheatsheet

### CODE REVIEW

File: packages/cursorless-vscode/src/ide/vscode/VscodeFileSystem.ts [hunk 1]
```diff
@@ -1,52 +1,15 @@
...
-import { stat } from "fs/promises";
-import { max } from "lodash";
+import { Disposable, FileSystem, PathChangeListener } from "@cursorless/common";
+import { RelativePattern, Uri, workspace } from "vscode";
 
 export class VscodeFileSystem implements FileSystem {
   watchDir(path: string, onDidChange: PathChangeListener): Disposable {
-    // Just poll for now; we can take advantage of VSCode's sophisticated
-    // watcher later. Note that we would need to do a version check, as VSCode
-    // file watcher is only available in more recent versions of VSCode.
```
Comment (auscompgeek) [inline packages/cursorless-vscode/src/ide/vscode/VscodeFileSystem.ts:14]: Do we remember what versions this was referring to?

I suppose the types must exist if the type checker is happy, so I'm sure we're not dropping support for any version 😄

File: packages/cursorless-vscode/src/ide/vscode/VscodeFileSystem.ts [hunk 2]
```diff
@@ -1,52 +1,15 @@
...
-      this.maxMtimeMs = maxMtime;
-      this.onDidChange();
-    }
-  }
-
-  dispose() {
-    clearInterval(this.timer);
+    // FIXME: Support globs?
+    const watcher = workspace.createFileSystemWatcher(
+      new RelativePattern(Uri.file(path), "**"),
```
Comment (AndreasArvidsson) [inline packages/cursorless-vscode/src/ide/vscode/VscodeFileSystem.ts:8]: You don't need to convert the string path to a uri. The first argument can just be a string.

File: packages/cursorless-vscode/src/ide/vscode/VscodeFileSystem.ts [hunk 3]
```diff
@@ -1,52 +1,15 @@
...
+    );
+    watcher.onDidChange(onDidChange);
```
Comment (AndreasArvidsson) [inline packages/cursorless-vscode/src/ide/vscode/VscodeFileSystem.ts:10]: We should probably handle these disposables as well

https://github.com/cursorless-dev/cursorless/blob/e2198dbd50faf9a38fe143916a285356ed14ef68/packages/cursorless-vscode/src/ide/vscode/hats/VscodeHatRenderer.ts#L419-L425

File: packages/cursorless-vscode/src/ide/vscode/VscodeFileSystem.ts (hunk 2 above)
Comment (pokey) [reply packages/cursorless-vscode/src/ide/vscode/VscodeFileSystem.ts:8]: All of the examples [in the docs](https://code.visualstudio.com/api/references/vscode-api#RelativePattern:~:text=Out%20of%20workspace%20file%20watching) use `Uri.file`, so I'd be inclined to follow their lead

File: packages/cursorless-vscode/src/ide/vscode/VscodeFileSystem.ts (hunk 3 above)
Comment (AndreasArvidsson) [reply packages/cursorless-vscode/src/ide/vscode/VscodeFileSystem.ts:10]: I couldn't find anything definitive so I would err on the side of caution.

File: packages/cursorless-vscode/src/ide/vscode/VscodeFileSystem.ts (hunk 1 above)
Comment (pokey) [reply packages/cursorless-vscode/src/ide/vscode/VscodeFileSystem.ts:14]: Good call; I've bumped our min version to the one that supports watching arbitrary files. Note that the type system doesn't help us here, because it's the same api but it would fail at runtime if you tried to watch a file outside the user's current project

**Comment by pokey** [comment]
I bumped min version in this PR to match the legacy version that we test (1.66.0), as that version is over 1 year old, and per https://github.com/cursorless-dev/cursorless/issues/1846, we're comfortable only supporting a year back, as users with older VSCode versions will just stop receiving Cursorless upgrades, which seems reasonable if they're on a very old VSCode version. Fwiw this PR requires version 1.64.0 to watch arbitrary paths

File: packages/cursorless-vscode/src/ide/vscode/VscodeFileSystem.ts (hunk 2 above)
Comment (pokey) [reply packages/cursorless-vscode/src/ide/vscode/VscodeFileSystem.ts:9]: huh interesting. I guess that's working so sure I'll change it here

File: packages/cursorless-vscode/src/ide/vscode/VscodeFileSystem.ts (hunk 3 above)
Comment (pokey) [reply packages/cursorless-vscode/src/ide/vscode/VscodeFileSystem.ts]: fwiw it could also backfire, because it's possible disposing one of these `onFoo` methods after disposing the main watcher could cause an error to be thrown. I wonder if we should at least dispose the watcher after the `onFoo` methods 🤔

Comment (pokey) [reply packages/cursorless-vscode/src/ide/vscode/VscodeFileSystem.ts]: k just went spelunking through VSCode code base, and the code uses an [`Emitter`](https://github.com/microsoft/vscode/blob/9e9c4e4bf713e143063414c59c8056b113f0263b/src/vs/base/common/event.ts#L946) for each event type (change / remove / add) and [disposes of its emitters](https://github.com/microsoft/vscode/blob/9e9c4e4bf713e143063414c59c8056b113f0263b/src/vs/workbench/api/common/extHostFileSystemEventService.ts#L104). From what I can tell there's no need to individually remove listeners when the emitter is disposed, and in the VSCode git extension, they only [dispose of the watcher](https://github.com/microsoft/vscode/blob/9e9c4e4bf713e143063414c59c8056b113f0263b/extensions/git/src/watch.ts#L19).

So long story short, we don't need to dispose of the watchers, and in fact I worry it could cause problems if we do dispose of them after the emitter is disposed

Comment (pokey) [reply packages/cursorless-vscode/src/ide/vscode/VscodeFileSystem.ts]: Assuming you have no issues with my conclusions here, I can fix your code while I'm here

Comment (AndreasArvidsson) [reply packages/cursorless-vscode/src/ide/vscode/VscodeFileSystem.ts:11]: Go ahead

**Comment by pokey** [comment]
@AndreasArvidsson this one is ready to go from my perspective

**Review by AndreasArvidsson — APPROVED** [review]
